pkg/pillar/persistcache/persistcache.go

```diff
@@ -0,0 +1,188 @@
+}
+
+// New values from cache or creates path if there's none
+func New(path string) (*persistCache, error) {
+	pc := &persistCache{}
+	pc.root = path
+	pc.cache = make(map[string]string)
+	pc.tempDir = filepath.Join(os.TempDir(), "persistCache")
+	if err := os.MkdirAll(pc.tempDir, FILE_MASK); err != nil {
+		return nil, err
+	}
+	defer pc.createLockFile()
+
+	if _, err := os.Stat(pc.root); os.IsNotExist(err) {
+		if err := os.MkdirAll(pc.root, FILE_MASK); err != nil {
+			return nil, err
+		}
+		return pc, nil
+	}
+
+	if _, err := os.Stat(filepath.Join(pc.tempDir, LOCK_FILE_NAME)); err == nil {
+		return nil, &LockFileExists{}
+	}
+
+	err := filepath.WalkDir(pc.root, func(path string, di fs.DirEntry, err error) error {
+		// We skip all directories
+		if di.IsDir() {
+			return nil
+		}
+
+		// lazy initialization
+		pc.cache[di.Name()] = ""
+
+		return nil
+	})
+
+	return pc, err
+}
+func (pc *persistCache) createLockFile() error {
+	lockfile, err := os.Create(filepath.Join(pc.tempDir, LOCK_FILE_NAME))
+	if err != nil {
+		return err
+	}
+	lockfile.Close()
+	return nil
+}
+
+// Close function is called to delete lock file so that other
+func (pc *persistCache) Close() error {
+	return os.Remove(filepath.Join(pc.tempDir, LOCK_FILE_NAME))
+}
+
+func (pc *persistCache) loadObject(objName string) error {
+	path := filepath.Join(pc.root, objName)
+	val, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	pc.cache[filepath.Base(path)] = string(val)
+
+	return nil
+}
+
+// Get value from cache
+func (pc *persistCache) Get(key string) (string, error) {
+	pc.Lock()
+	defer pc.Unlock()
+
+	if _, err := os.Stat(filepath.Join(pc.tempDir, LOCK_FILE_NAME)); os.IsNotExist(err) {
+		return "", &LockFileNotFound{}
+	}
+
+	if pc.cache[key] == "" {
+		pc.loadObject(key)
+	}
+
+	val, _ := pc.cache[key]
+
+	return val, nil
+}
+
+// Create or update value in in-memory cache and filesystem
+func (pc *persistCache) Put(key string, val string) (string, error) {
+	pc.Lock()
+	defer pc.Unlock()
+
+	if _, err := os.Stat(filepath.Join(pc.tempDir, LOCK_FILE_NAME)); os.IsNotExist(err) {
+		return "", &LockFileNotFound{}
+	}
+
+	if !isValidKey(key) {
+		return "", &InvalidKeyError{}
+	}
+	if !isValidValue(val) {
+		return "", &InvalidValueError{}
+	}
+
+	// save file
+	filepath := filepath.Join(pc.root, key)
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, FILE_MASK)
+	if err != nil {
+		return "", err
+	}
+	f.WriteString(val)
+	if err := f.Close(); err != nil {
+		return "", err
+	}
+
+	pc.cache[key] = val
+
+	return filepath, nil
+}
+
+// Remove element from cache and filesystem
+func (pc *persistCache) Delete(key string) error {
+	pc.Lock()
+	defer pc.Unlock()
+
+	if _, err := os.Stat(filepath.Join(pc.tempDir, LOCK_FILE_NAME)); os.IsNotExist(err) {
+		return &LockFileNotFound{}
+	}
+
+	delete(pc.cache, key)
+	return os.Remove(filepath.Join(pc.root, key))
+}
+
+func isValidKey(key string) bool {
+	if strings.Contains(key, "/") {
```

Comment:
What about names like ".." and "."?
I think it makes sense to call path.Clean() on the key (since it will replace .. with ../ etc).
Then you can check for "/".

Reply:
Makes sense, added it